### PR TITLE
Bucket Door Prank

### PR DESCRIPTION
**Pull request by amylizzle**
## About the PR 

Enabling whole new levels of clown antics, you can now place a bucket above a door so the next person walking through gets a bucket on their head!
Fill the bucket with ~~fluorosulphuric acid~~ glitter for extra pranking!

https://user-images.githubusercontent.com/3855802/209445513-59dfa1df-9176-4e7e-bf12-5fbcc3489d16.mp4

https://user-images.githubusercontent.com/3855802/209445516-7b006740-f081-4a6f-97c6-b7392e4525a4.mp4

If you're quick enough, or observant enough, you can escape the bucket pranking by triggering the door and moving away. Pranked doors get a little bucket added to their icon so you can see it.

## Why's this needed? 
Funny.

## Changelog 

```changelog
(u)Amylizzle
(*)Prank your friends and/or enemies with a hilarious bucket-above-the-door prank! Just click a wooden door with a bucket to set it up. Don't forget to fill the bucket with glitter for some extra harmless fun!
```

### CODE REVIEW

File: code/modules/chemistry/Chemistry-Tools.dm [hunk 1]
```diff
@@ -508,6 +508,68 @@ proc/ui_describe_reagents(atom/A)
...
+			return ..()
+		if(locate(/obj/item/reagent_containers/glass/bucket) in target)
+			boutput(user,"<b>There's already a bucket prank set up!</b>")
+			return ..()
+		boutput(user, "You start propping \the [src] above \the [target]...")
+		SETUP_GENERIC_ACTIONBAR(user, src, 3 SECONDS, .proc/setup_bucket_prank, list(target, user), src.icon, src.icon_state, \
+					src.visible_message("<span class='alert'><B>[user] props a [src] above \the [target]</B></span>"), \
+					INTERRUPT_ACT | INTERRUPT_STUNNED | INTERRUPT_ACTION | INTERRUPT_MOVE)
+
+	proc/setup_bucket_prank(obj/machinery/door/D, mob/user)
```
Comment (TobleroneSwordfish) [inline code/modules/chemistry/Chemistry-Tools.dm:522]: non-standard single letter var 💢

File: code/modules/chemistry/Chemistry-Tools.dm [hunk 2]
```diff
@@ -508,6 +508,68 @@ proc/ui_describe_reagents(atom/A)
...
+			src.visible_message("<span class='alert'>[src] falls from \the [D][splash? ", splashing its contents on the floor" : ""].</span>")
+			return
+		else //we're in range, splash the AM, splash the floor
+			logTheThing(LOG_COMBAT, AM, "Victim of bucket-door-prank with reagents: [log_reagents(src)] on [D]")
+			src.reagents.reaction(AM, TOUCH, src.reagents.total_volume/2) //half on the mover
+			src.reagents.reaction(get_turf(D)) //half on the floor
+			if(ishuman(AM))
+				//the bucket lands on your head for maximum comedy
+				var/mob/living/carbon/human/H = AM
+				var/obj/item/clothing/head/helmet/bucket/hat/bucket_hat = new hat_bucket_type(src.loc)
```
Comment (TobleroneSwordfish) [inline code/modules/chemistry/Chemistry-Tools.dm:554]:
```suggestion
				var/obj/item/clothing/head/helmet/bucket/hat/bucket_hat = new src.hat_bucket_type(src.loc)
```

File: code/modules/chemistry/Chemistry-Tools.dm [hunk 3]
```diff
@@ -508,6 +508,68 @@ proc/ui_describe_reagents(atom/A)
...
+		//note that AM can be null, and this is caught by !IN_RANGE
+		UnregisterSignal(D, COMSIG_DOOR_OPENED_BY)
+		D.UpdateOverlays(null, "bucketprank")
+		var/splash = (src.reagents.total_volume > 1)
+		if(!IN_RANGE(AM, D, 1))
+			src.set_loc(get_turf(D))
+			src.reagents.reaction(get_turf(D))
+			src.visible_message("<span class='alert'>[src] falls from \the [D][splash? ", splashing its contents on the floor" : ""].</span>")
+			return
+		else //we're in range, splash the AM, splash the floor
```
Comment (TobleroneSwordfish) [inline code/modules/chemistry/Chemistry-Tools.dm:547]: else not needed since the other case returns

File: code/obj/machinery/door/door_parent.dm [hunk 4]
```diff
@@ -367,6 +370,7 @@
 	if (src.allowed(AM))
 		if (src.density)
 			last_used = world.time
+			bumper = AM
```
Comment (TobleroneSwordfish) [inline code/obj/machinery/door/door_parent.dm:373]:
```suggestion
			src.bumper = AM
```

File: code/modules/chemistry/Chemistry-Tools.dm [hunk 5]
```diff
@@ -508,6 +508,68 @@ proc/ui_describe_reagents(atom/A)
...
+			if(ishuman(AM))
+				//the bucket lands on your head for maximum comedy
+				var/mob/living/carbon/human/H = AM
+				var/obj/item/clothing/head/helmet/bucket/hat/bucket_hat = new hat_bucket_type(src.loc)
+				if(isnull(H.head))
+					H.equip_if_possible(bucket_hat, H.slot_head)
+					H.set_clothing_icon_dirty()
+					H.visible_message("<span class='alert'>[src] falls from \the [D], landing on [H] like a hat[splash? ", and splashing [him_or_her(H)] with its contents" : ""]! [pick("Peak comedy!","Hilarious!","What a tool!")]</span>", \
+										"<span class='alert'>[src] falls from \the [D], landing on your head like a hat[splash? ", and splashing you with its contents" : ""]!</span>")
+					qdel(src) //it's a hat now
```
Comment (TobleroneSwordfish) [inline code/modules/chemistry/Chemistry-Tools.dm:560]: exists in both cases so lift out of condition

File: _std/defines/component_defines/component_defines_special.dm [hunk 6]
```diff
@@ -132,3 +132,7 @@
 #define COMSIG_LIGHT_ENABLE "enable_light"
 /// Send to a thing to disable component lights on it
 #define COMSIG_LIGHT_DISABLE "disable_light"
+
+// ---- Door signals, for bucket pranks ----
+/// When the door was bumped open, send the movable that opened it
+#define COMSIG_DOOR_OPENED_BY "door_opened_by"
```
Comment (TobleroneSwordfish) [inline _std/defines/component_defines/component_defines_special.dm:138]: I think most signal names omit the "by" part

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
No idea if this is balanced but it's funny.

File: code/modules/chemistry/Chemistry-Tools.dm (hunk 1 above)
Comment (amylizzle) [reply code/modules/chemistry/Chemistry-Tools.dm:522]: but I had so much writing `\the [D]`!

File: code/modules/chemistry/Chemistry-Tools.dm (hunk 3 above)
Comment (amylizzle) [reply code/modules/chemistry/Chemistry-Tools.dm:547]: Not really needed, but I think it makes it clearer. I might just ditch the `return` instead.

**Comment by frawhst** [comment]
This is funny so I like it
Seems weird to me that the bucket is off to the side of the door, and I'm personally fine with just putting it """on top""" (probably just in the middle for E/W doors) of the door. Communicating mechanics is more important than making logical sense, probably.

**Comment by Tarmunora** [comment]
self-quoting from discord discussion re this:
> :lazy: I could see it for like, the ye olde wooden doors
but not for airlocks
> eh, it's as much the fact that booby-trapping a pressure-sealed airlock by... propping up a plastic bucket(?) just feels off

**Comment by Sord213** [comment]
clown leaving a bucket full of lube on security's door is peak comedy

**Comment by amylizzle** [comment]
I tried bucket-in-the-middle first and it looked not great. Mostly because it ended up looking the same as if someone had just dropped an item on the floor and the door had closed. Before the I applied the scale to the bucket, it was visually indistinguishable from someone having just dropped a bucket in a doorway.

My concern with the wooden doors only approach is that there are so few of those, and while it would kinda make more sense, it's also a lot less fun. It would be a pretty trivial change to make though, should it become necessary. You'd literally just need to change one `istype()` call on the bucket.

I do hope it gets merged as is. I realise there are genuine concerns regarding balance, but with the logging and general anti-griefing rules that exist, I *think* it'd be fine.

**Comment by frawhst** [comment]
My thing's addressed, if middle-of-door looks bad then I don't mind 👍

**Comment by colossusqw** [comment]
This is genius and i love it.

**Comment by TobleroneSwordfish** [comment]
Okay after some deliberation and the discovery that you can actually build wooden doors, we're going to say this is good to merge if restricted to wooden doors. (sorry it took a while, I was planning on adding the ability to make wooden doors but it turns out you can already)

**Pull request merged**